### PR TITLE
Refactor MetaData to split off the concrete index name logic to a dedicated service

### PR DESCRIPTION
**Pull request by martijnvg**
Changes in a nutshell:
- All expression logic is now encapsulated by `ExpressionResolver` interface.
- `MetaData#convertFromWildcards()` gets replaced by `WildcardExpressionResolver`.
- All of the indices expansion methods are being moved from `MetaData` class to the new `IndexNameExpressionResolver` class.
- All single index expansion optimisations are removed.

The logic for resolving a concrete index name from an expression has been moved from the MetaData class to IndexNameExpressionResolver class. The logic has been cleaned up and simplified were was possible without breaking bwc. Mainly the logic that deals if there is just one index expression is no more, mainly to make to code simpler. Also the notion of aliasOrIndex has been changed to index expression.

The IndexNameExpressionResolver translates index name expressions into concrete indices. The list of index name expressions are first delegated to the known IndexExpressionFilter. An IndexExpressionFilter is responsible for translating if possible an expression into a concrete indices or aliases otherwise the expression is left untouched. Concretely this means converting wildcard expressions into concrete indices or aliases, but in the future other implementations could convert expressions based on different rules.

Beyond this there is also some follow up work to do:
- Replace SnapshotUtils#filterIndices by using the IndexNameExpressionResolver
- During the lifetime of a request index expressions or resolved to concrete indices multiple times. (finding the search routing, whether aliases have filters, checking for cluster blocks and just resolving the index name expressions to concrete indices) This can be reduced to a single lookup. 
- Revise the in memory data structures used by MetaData. This is interesting in the case of many indices and aliases. For example I think that `MetaData#aliasAndIndexToIndexMap` can be removed in order to save heap memory at the cost of some extra look ups at request time. Also a completely different approach can be taken to store index and aliases with their respective IndexMetaData and AliasMetaData into a FST.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java [hunk 1]
```diff
@@ -0,0 +1,610 @@
...
+        IndicesOptions options = context.getOptions();
+        boolean failClosed = options.forbidClosedIndices() && options.ignoreUnavailable() == false;
+        boolean failNoIndices = options.ignoreUnavailable() == false;
+
+        // TODO: this exception in the logic must be revised:
+        // I think this is an inconsistency (bug) in the old single expressions logic. (there was different logic for a single and multiple index logic)
+        // This must be addressed in a follow up issue
+        if (indexExpressions.length == 1) {
+            failNoIndices = options.allowNoIndices() == false;
+        }
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: I did quite some digging around this special case. It is there from the early days of indices options and back when I added unit tests for it to MetaDataTests I probably just added tests to verify the current behaviour (assuming that what we had was ok). Looking at it more I don't get this test that fails when you remove this special case:

```
IndicesOptions noExpandErrorUnavailable = IndicesOptions.fromOptions(false, true, false, false);

            String[] results = md.concreteIndices(noExpandErrorUnavailable, "baz*");
            assertThat(results, emptyArray());

            try {
                md.concreteIndices(noExpandErrorUnavailable, "foo", "baz*");
                fail();
            } catch (IndexMissingException e) {
                assertThat(e.index().name(), equalTo("baz*"));
            }
```

it makes no sense to me that with same options, a different number of indices leads to a different result... I think we could adapt the test and remove this special case (in a separate issue it's fine too)

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java [hunk 2]
```diff
@@ -0,0 +1,610 @@
...
+import org.elasticsearch.indices.IndexMissingException;
+
+import java.util.*;
+
+import static com.google.common.collect.Lists.newArrayList;
+import static com.google.common.collect.Maps.newHashMap;
+
+public class IndexNameExpressionResolver {
+
+    private final ImmutableList<IndexExpressionFilter> parsers;
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: not sure I would call this `parsers`, maybe `expressionFilters` ?

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java [hunk 3]
```diff
@@ -0,0 +1,610 @@
...
+     * @throws IndexMissingException    if the resolved index or alias provided doesn't exist
+     * @throws IllegalArgumentException if the index resolution lead to more than one index
+     * @return the concrete index obtained as a result of the index resolution
+     */
+    public String concreteSingleIndex(ClusterState state, IndicesRequest request) throws IndexMissingException, IllegalArgumentException {
+        String index = request.indices() != null && request.indices().length > 0 ? request.indices()[0] : null;
+        return concreteSingleIndex(state, index, request.indicesOptions());
+    }
+
+    public String concreteSingleIndex(ClusterState state, DocumentRequest request) throws IndexMissingException, IllegalArgumentException {
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: I think DocumentRequest could potentially extend IndicesRequest, then we would not need this specific method. maybe a different PR though?

Comment (javanna) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: btw javadocs would be nice to have if we keep this public method

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java [hunk 4]
```diff
@@ -0,0 +1,610 @@
...
+    /**
+     * Iterates through the list of indices and selects the effective list of filtering aliases for the
+     * given index.
+     * <p/>
+     * <p>Only aliases with filters are returned. If the indices list contains a non-filtering reference to
+     * the index itself - null is returned. Returns <tt>null</tt> if no filtering is required.</p>
+     */
+    // TODO: merge this logic with concreteIndices()
+    // This will allow any transport action to lookup the concrete index names, routing and whether alias have filters in one lookup,
+    // instead of multiple.
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: can we remove the TODO from code and open an issue instead?

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java [hunk 5]
```diff
@@ -0,0 +1,610 @@
...
+        }
+        if (filteringAliases == null) {
+            return null;
+        }
+        return filteringAliases.toArray(new String[filteringAliases.size()]);
+    }
+
+    // TODO: merge this logic with concreteIndices()
+    // This will allow any transport action to lookup the concrete index names, routing and whether alias have filters in one lookup,
+    // instead of multiple.
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: same as above about losing the TODO but opening an issue instead

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java [hunk 6]
```diff
@@ -0,0 +1,610 @@
...
+            for (String indexOrAlias : indicesOrAliases) {
+                if (Regex.isSimpleMatchPattern(indexOrAlias)) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    public static class Context {
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: does this beed to be public? maybe package private?

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java [hunk 7]
```diff
@@ -0,0 +1,610 @@
...
+    }
+
+    private interface IndexExpressionFilter {
+
+        /**
+         * Expands or changes the list of expressions in concrete indices and aliases
+         *
+         * @return a new list with expressions based on the provided expressions
+         */
+        List<String> operate(Context context, List<String> expressions);
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: not a big fan of calling it `operate`. If it's a filter maybe the method should be called `filter`. That said when I look at it I don't know anymore if it's really a filter, more of a resolver?

File: core/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java [hunk 8]
```diff
@@ -471,6 +473,8 @@ public boolean apply(IndexWarmersMetaData.Entry warmer) {
     /**
      * Returns indexing routing for the given index.
      */
+    // TODO: This can be moved to IndexNameExpressionResolver too, but this means that we will support wildcards and other expressions
+    // in the index,bulk,update and delete apis.
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java]: wouldn't it be possible to move this method where it belongs without supporting wildcards against it though? those apis shouldn't support wildcards

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java [hunk 9]
```diff
@@ -0,0 +1,610 @@
...
+    public IndexNameExpressionResolver() {
+        parsers = ImmutableList.<IndexExpressionFilter>of(new WildcardExpressionFilter());
+    }
+
+    public String[] concreteIndices(ClusterState state, IndicesRequest request) throws IndexMissingException, IllegalArgumentException {
+        Context context = new Context(state, request.indicesOptions());
+        return concreteIndices(context, request.indices());
+    }
+
+    public String[] concreteIndices(ClusterState state, IndicesOptions options, String... indexExpressions) throws IndexMissingException, IllegalArgumentException {
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: add some javadocs to these two public methods?

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java [hunk 10]
```diff
@@ -0,0 +1,610 @@
...
+import static com.google.common.collect.Lists.newArrayList;
+import static com.google.common.collect.Maps.newHashMap;
+
+public class IndexNameExpressionResolver {
+
+    private final ImmutableList<IndexExpressionFilter> parsers;
+
+    @Inject
+    public IndexNameExpressionResolver() {
+        parsers = ImmutableList.<IndexExpressionFilter>of(new WildcardExpressionFilter());
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: I wonder if the filter should be injected instead, and maybe even live in its own class file and be public. For instance plugins might want to have access to how we resolve wildcards without having to call concreteIndices. Either we have a method here to expose that logic or we make the WildcardExpressionFilter accessible from the outside.

**Comment by javanna** [comment]
I did a first review round, the change looks good, left a few comments. 

It is not the easiest PR to review given that we are moving whole methods and slightly changing them too meanwhile. I think that the most important aspects are that `convertFromWildcards` gets replaced by `WildcardExpressionFilter` and we potentially support for more filters that will come in the future. All of the indices expansion methods are being moved from MetaData to the new IndexNameExpressionResolver dedicated class, and meanwhile we removed all of the optimizations around single index usecase etc. anything else that might have missed?

**Comment by martijnvg** [comment]
thanks @javanna! and those changes are correct! (nothing missed)

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java (hunk 1 above)
Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: Ok, I'll change it via this PR.

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java (hunk 2 above)
Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: +1 (the initial name was something with parser :) )

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java (hunk 3 above)
Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: I'll try to this in this PR, if that change gets too large, I'll do it in a different PR.

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java (hunk 4 above)
Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: ++

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java (hunk 6 above)
Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: makes sense for now

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java (hunk 10 above)
Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: maybe for now we should just expose a method? Lets we have multiple expression filters, the order in which they are applied is important. (for example date math expressions should be resolved before wildcards are) So I'm a bit hesitant with opening IndexExpressionFilter to plugins yet.

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java (hunk 7 above)
Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: maybe we should name it `convert` and then rename `IndexExpressionFilter` to `IndexExpressionConverter`?

Comment (javanna) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: you against IndexExpressionResolver? isn't resolving what it does?

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java (hunk 10 above)
Comment (javanna) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: thinking more about it, I think it's fine as-is :)

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java (hunk 7 above)
Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: not in particular, maybe then we should call the method `resolve` :)

Comment (javanna) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: +1 to that if you like it too

File: core/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java (hunk 8 above)
Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java]: I tried to move it, but it is relative large change on top of this PR. Can this be done in a different PR?

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java (hunk 4 above)
Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: done: https://github.com/elastic/elasticsearch/issues/12173

**Comment by martijnvg** [comment]
@javanna I updated the PR.

File: core/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java (hunk 8 above)
Comment (javanna) [inline core/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java]: sure let's just not forget :)

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java [hunk 11]
```diff
@@ -0,0 +1,600 @@
...
+        public ClusterState getState() {
+            return state;
+        }
+
+        public IndicesOptions getOptions() {
+            return options;
+        }
+    }
+
+    private interface IndexExpressionFilter {
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: I thought we said we would rename this to IndexExpressionResolver, maybe that was not clear though. what do you think about that?

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java [hunk 12]
```diff
@@ -0,0 +1,600 @@
...
+         * @return a new list with expressions based on the provided expressions
+         */
+        List<String> resolve(Context context, List<String> expressions);
+
+    }
+
+    /**
+     * Expends of changes expressions with wildcards into concrete indices and aliases
+     */
+    final static class WildcardExpressionFilter implements IndexExpressionFilter {
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: same as above, WildcardExpressionResolver?

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java [hunk 13]
```diff
@@ -0,0 +1,600 @@
...
+import org.elasticsearch.indices.IndexMissingException;
+
+import java.util.*;
+
+import static com.google.common.collect.Lists.newArrayList;
+import static com.google.common.collect.Maps.newHashMap;
+
+public class IndexNameExpressionResolver {
+
+    private final ImmutableList<IndexExpressionFilter> expressionFilters;
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: if we rename to IndexExpressionResolver then expressionFilters => expressionResolvers?

**Comment by javanna** [comment]
left a couple more comments

**Comment by martijnvg** [comment]
@javanna I updated the PR.

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java [hunk 14]
```diff
@@ -0,0 +1,599 @@
...
+                }
+                String[] indices;
+                if (options.expandWildcardsOpen() && options.expandWildcardsClosed()) {
+                    indices = metaData.concreteAllIndices();
+                } else if (options.expandWildcardsOpen()) {
+                    indices = metaData.concreteAllOpenIndices();
+                } else if (options.expandWildcardsClosed()) {
+                    indices = metaData.concreteAllClosedIndices();
+                } else {
+                    assert false : "convertFromWildcards shouldn't get called if wildcards expansion is disabled";
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: update the assert message given that convertFromWIldcards is not a thing anymore?

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java [hunk 15]
```diff
@@ -0,0 +1,599 @@
...
+         * Expands or changes the list of expressions in concrete indices and aliases
+         *
+         * @return a new list with expressions based on the provided expressions
+         */
+        List<String> resolve(Context context, List<String> expressions);
+
+    }
+
+    /**
+     * Expends of changes expressions with wildcards into concrete indices and aliases
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: repalce with "Resolver for index name expression that contain wildcards, which will rsolve them into their corresponding concrete indices or aliases" ?

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java [hunk 16]
```diff
@@ -0,0 +1,599 @@
...
+
+        public IndicesOptions getOptions() {
+            return options;
+        }
+    }
+
+    private interface IndexExpressionResolver {
+
+        /**
+         * Expands or changes the list of expressions in concrete indices and aliases
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: Resolves rather then expands or changes ?

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java [hunk 17]
```diff
@@ -0,0 +1,599 @@
...
+            this.options = options;
+        }
+
+        public ClusterState getState() {
+            return state;
+        }
+
+        public IndicesOptions getOptions() {
+            return options;
+        }
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: minor nitpick, maybe we can even lose the getters here given that this is package private and the fields are final?

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java [hunk 18]
```diff
@@ -0,0 +1,599 @@
...
+                }
+            }
+        }
+        if (filteringAliases == null) {
+            return null;
+        }
+        return filteringAliases.toArray(new String[filteringAliases.size()]);
+    }
+
+    public Map<String, Set<String>> resolveSearchRouting(ClusterState state, @Nullable String routing, String... expressions) {
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: add some javadocs? not sure if it had it before though :)

File: core/src/test/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolverTests.java [hunk 19]
```diff
@@ -404,139 +427,101 @@ public void testIndexOptions_singleIndexNoExpandWildcards() {
...
-                .put(indexBuilder("testXYY").state(State.CLOSE))
-                .put(indexBuilder("testYYY").state(State.OPEN))
-                .put(indexBuilder("testYYX").state(State.CLOSE))
-                .put(indexBuilder("kuku").state(State.OPEN));
-        MetaData md = mdBuilder.build();
-        // Can't test when wildcard expansion is turned off here as convertFromWildcards shouldn't be called in this case.  Tests for this are covered in the concreteIndices() tests
-        assertThat(newHashSet(md.convertFromWildcards(new String[]{"testX*"}, IndicesOptions.fromOptions(true, true, true, true))), equalTo(newHashSet("testXXX", "testXXY", "testXYY")));
-        assertThat(newHashSet(md.convertFromWildcards(new String[]{"testX*"}, IndicesOptions.fromOptions(true, true, false, true))), equalTo(newHashSet("testXYY")));
-        assertThat(newHashSet(md.convertFromWildcards(new String[]{"testX*"}, IndicesOptions.fromOptions(true, true, true, false))), equalTo(newHashSet("testXXX", "testXXY")));
-    }
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolverTests.java]: where did these tests go?

**Comment by javanna** [comment]
left a bunch of minor comments, LGTM otherwise

File: core/src/test/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolverTests.java (hunk 19 above)
Comment (martijnvg) [inline core/src/test/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolverTests.java]: it is in WildcardExpressionResolverTests

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java (hunk 17 above)
Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: I prefer to keep the getters, if that is okay?

Comment (javanna) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: sure

File: core/src/test/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolverTests.java (hunk 19 above)
Comment (javanna) [inline core/src/test/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolverTests.java]: good
